extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/HibernateOrmRecorder.java

```diff
@@ -105,8 +106,16 @@ public Supplier<JPAConfig> jpaConfigSupplier(HibernateOrmRuntimeConfig config) {
         return () -> new JPAConfig(config);
     }
 
-    public void startAllPersistenceUnits(BeanContainer beanContainer) {
-        beanContainer.beanInstance(JPAConfig.class).startAll();
+    public void startAllPersistenceUnits(BeanContainer beanContainer, ShutdownContext shutdownContext) {
+        JPAConfig jpaConfig = beanContainer.beanInstance(JPAConfig.class);
+        // NOTE:
+        //  - We register the shutdown task before we start any PUs,
+        //    This way we'll be able to clean up even if one of the PUs fails to start while others already did.
+        //  - The step that starts the PUs returns the ServiceStartBuildItem, this in turn ensures that
+        //    the shutdown task that triggers the ShutdownEvent will be registered after this one,
+        //    and users will have access to the "ORM stuff" in their listeners.
+        shutdownContext.addShutdownTask(jpaConfig::shutdown);
```

Comment:
Do you know what guarantees that datasources are stopped after JPA exactly?

Reply:
mmm so the datasources we use here are coming from CDI, right ? 
and since CDI container starts (and adds it's shutdown task) before ORM does, all beans are still around. So as long as the datasource does not have it's own shutdown we should be ok. 
Also we depend on `List<JdbcDataSourceBuildItem> dataSourcesConfigured` in `startPersistenceUnits` so that means the bean producer was created. And there are no shutdown tasks in agroal (at least none that I could find).

Reply:
Ok, so you're saying the shutdown order is this?

1. CDI shutdown event observers
2. JPAConfig shutdown
3. CDI beans destruction (including agroal)

Reply:
well it's probably "Shutdown event is fired and all observers process the event" (1)

but yes, that's the order 🙂